File: BinaryTree and Divide and Conquer/le114_flattenBinaryTreeToLinkedList.py
"""
Given a binary tree, flatten it to a linked list in-place.

For example,
Given

         1
        / \
       2   5
      / \   \
     3   4   6
The flattened tree should look like:
   1
    \
     2
      \
       3
        \
         4
          \
           5
            \
             6


"""

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def flatten(self, root):
        """
        :type root: TreeNode
        :rtype: void Do not return anything, modify root in-place instead.
        """
        if not root:
            return
        self.lastNode = None
        self.dfsHelper(root)

        
    def dfsHelper(self, root): # i am not familiar with this case when reference is a node not a list/dict
        if not root:
            return
        if self.lastNode:
            self.lastNode.left = None
            self.lastNode.right = root
        self.lastNode = root
        right = root.right
        self.dfsHelper(root.left) # after this step root's location (direction) has been changed
        self.dfsHelper(right)

File: JavaPartI/le114_flattenBinaryTreeToLinkedList.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public void flatten(TreeNode root) {
        flat(root);
    }
    
    private TreeNode flat(TreeNode root) {
        if (root == null) {
            return root;
        }
        TreeNode tail1 = flat(root.left);
        TreeNode tail2 = flat(root.right);
        if (tail1 != null) {
            tail1.right = root.right;
            root.right = root.left;
            root.left = null;
        } 
        if (tail2 != null) {
            return tail2;
        } 
        if (tail1 != null) {
            return tail1;
        }
        return root;
    } 
}
/// preorder traversal
class Solution {
    private TreeNode prev = null;
    public void flatten(TreeNode root) {
        if (root == null) {
            return;
        }
        if (prev != null) {
            prev.left = null;
            prev.right = root;
        }
        prev = root;
        TreeNode right = root.right;// root.right will be changed after using flatten(root.let). left node will be conencted to the right
        flatten(root.left);
        flatten(right);
    }
}
